transport/src/main/java/io/netty/channel/oio/AbstractOioMessageChannel.java

```diff
@@ -89,13 +96,9 @@ protected void doRead() {
             if (isOpen()) {
                 unsafe().close(unsafe().voidPromise());
             }
-        } else if (allocHandle.lastBytesRead() == 0 && isActive()) {
-            // If the read amount was 0 and the channel is still active we need to trigger a new read()
-            // as otherwise we will never try to read again and the user will never know.
-            // Just call read() is ok here as it will be submitted to the EventLoop as a task and so we are
-            // able to process the rest of the tasks in the queue first.
-            //
-            // See https://github.com/netty/netty/issues/2404
+        } else if (readPending || config.isAutoRead() || !readData && isActive()) {
```

Comment:
humm scary claim ... and even scarier this seems like a false positive.

- readPending is expected to be set to true in fireChannelRead() or fireChannelReadComplete()
- config.isAutoRead() is independent of this entire code block and can also be set by the user in the above methods
- readData may be false if nothing was read, isActive() is true if the channel has not been closed (due to read error or the user explicitly closed it)